### PR TITLE
ImageBitmapBacking is a redundant class

### PR DESCRIPTION
**Pull request by kkinnunen-apple**
#### 44884327f7156a080986bb7f9748e825347a3684
<pre>
ImageBitmapBacking is a redundant class
<a href="https://bugs.webkit.org/show_bug.cgi?id=267455">https://bugs.webkit.org/show_bug.cgi?id=267455</a>
<a href="https://rdar.apple.com/120906757">rdar://120906757</a>

Reviewed by Matt Woodrow.

ImageBitmap would hold a ImageBitmapBacking. It is a bit error-prone,
as the purpose of the backing is not much more than just hold the
ImageBuffer. Backing might get nulled, as might the ImageBuffer itself.

Make it simpler and move the ImageBitmapBacking directly to ImageBitmap.
This also removes the redundant allocation of ImageBitmapBacking.

Make the Web Worker serialization work similarly to OffscreenCanvas:
when the ImageBitmap is &quot;detached&quot; for moving to a Worker, construct
a DetachedImageBitmap to move it.

* Source/WebCore/Headers.cmake:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/bindings/js/SerializedScriptValue.cpp:
(WebCore::CloneSerializer::dumpImageBitmap):
(WebCore::CloneDeserializer::deserialize):
(WebCore::CloneDeserializer::CloneDeserializer):
(WebCore::CloneDeserializer::readTransferredImageBitmap):
(WebCore::CloneDeserializer::readImageBitmap):
(WebCore::SerializedScriptValue::SerializedScriptValue):
(WebCore::SerializedScriptValue::computeMemoryCost const):
(WebCore::SerializedScriptValue::create):
(WebCore::SerializedScriptValue::deserialize):
* Source/WebCore/bindings/js/SerializedScriptValue.h:
(WebCore::SerializedScriptValue::SerializedScriptValue):
* Source/WebCore/html/ImageBitmap.cpp:
(WebCore::DetachedImageBitmap::DetachedImageBitmap):
(WebCore::ImageBitmap::create):
(WebCore::ImageBitmap::detach):
(WebCore::ImageBitmap::createBlankImageBuffer):
(WebCore::ImageBitmap::createCompletionHandler):
(WebCore::ImageBitmap::createFromBuffer):
(WebCore::ImageBitmap::ImageBitmap):
(WebCore::ImageBitmap::takeImageBuffer):
(WebCore::ImageBitmap::width const):
(WebCore::ImageBitmap::height const):
(WebCore::ImageBitmap::updateMemoryCost):
(WebCore::ImageBitmap::~ImageBitmap): Deleted.
(WebCore::ImageBitmap::takeImageBitmapBacking): Deleted.
* Source/WebCore/html/ImageBitmap.h:
(WebCore::DetachedImageBitmap::memoryCost const):
* Source/WebCore/html/ImageBitmapBacking.cpp: Removed.
* Source/WebCore/html/ImageBitmapBacking.h: Removed.
* Source/WebCore/html/OffscreenCanvas.cpp:
(WebCore::OffscreenCanvas::transferToImageBitmap):
* Source/WebCore/platform/graphics/ImageBuffer.cpp:
* Source/WebCore/platform/graphics/ImageBuffer.h:
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:
* Source/WebKit/WebProcess/GPU/graphics/RemoteImageBufferProxy.h:

Canonical link: <a href="https://commits.webkit.org/273213@main">https://commits.webkit.org/273213@main</a>
</pre>

https://github.com/WebKit/WebKit/commit/b6f9c78a1540fd22251921c60c5030bf4457916d

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/34566 "4 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/13390 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/36573 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/37256 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/31239 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/35708 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/15788 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/10497 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/30217 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/35091 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/11328 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/30802 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/9899 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/9983 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/30866 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/38534 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/31390 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/31182 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/36045 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/10099 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/7991 "Passed tests") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/33998 "Found 1 new API test failure: /WebKitGTK/TestWebKitWebContext:/webkit/WebKitWebContext/proxy (failure)") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/11922 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/7963 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/10652 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/10970 "Built successfully") | | |

### CODE REVIEW

File: Source/WebCore/html/ImageBitmap.cpp [hunk 1]
```diff
@@ -125,12 +150,16 @@ RefPtr<ImageBuffer> ImageBitmap::createImageBuffer(ScriptExecutionContext& scrip
...
 }
 
-std::optional<ImageBitmapBacking> ImageBitmap::detach()
+std::optional<DetachedImageBitmap> ImageBitmap::detach()
 {
-    std::optional<ImageBitmapBacking> backing = takeImageBitmapBacking();
-    if (backing)
-        backing->disconnect();
-    return backing;
+    if (m_bitmap) {
```
Comment (shallawa) [inline Source/WebCore/html/ImageBitmap.cpp]: I think an early return here will make the code clearer.

File: Source/WebCore/html/ImageBitmap.h [hunk 2]
```diff
@@ -66,6 +67,21 @@ struct ImageBitmapOptions;
 
 template<typename IDLType> class DOMPromiseDeferred;
 
+class DetachedImageBitmap {
+public:
+    DetachedImageBitmap(DetachedImageBitmap&&);
+    WEBCORE_EXPORT ~DetachedImageBitmap();
+    DetachedImageBitmap& operator=(DetachedImageBitmap&&);
+    size_t memoryCost() const { return m_bitmap->memoryCost(); }
```
Comment (shallawa) [inline Source/WebCore/html/ImageBitmap.h:75]: Maybe adding a function here to return an `OptionSet<ImageBitmapSerializationFlags>` can make the code in SerializedScriptValue.cpp a littler simpler.

File: Source/WebCore/html/ImageBitmap.h [hunk 3]
```diff
@@ -152,8 +164,11 @@ class ImageBitmap final : public ScriptWrappable, public RefCounted<ImageBitmap>
     static void createFromBuffer(ScriptExecutionContext&, Ref<ArrayBuffer>&&, String mimeType, long long expectedContentLength, const URL&, ImageBitmapOptions&&, std::optional<IntRect>, ImageBitmapCompletionHandler&&);
     void updateMemoryCost();
 
-    std::optional<ImageBitmapBacking> m_backingStore;
+    RefPtr<ImageBuffer> m_bitmap;
     std::atomic<size_t> m_memoryCost { 0 };
+    const bool m_originClean : 1 { false };
+    const bool m_premultiplyAlpha : 1 { false };
+    const bool m_forciblyPremultiplyAlpha : 1 { false };
```
Comment (shallawa) [inline Source/WebCore/html/ImageBitmap.h:171]: It is unfortunate to repeat the same flags here and in `DetachedImageBitmap`

File: Source/WebCore/html/ImageBitmap.h (hunk 2 above)
Comment (kkinnunen-apple) [reply Source/WebCore/html/ImageBitmap.h:75]: The detachedimagebitmap is not used to serialize a out-of-process transfer of imagebitmap. It's only used for in-memory transfer..

File: Source/WebCore/html/ImageBitmap.h (hunk 3 above)
Comment (kkinnunen-apple) [reply Source/WebCore/html/ImageBitmap.h:171]: I'm not sure how much of a problem it is,  having these duplicated in the DetachedImageBitmap and here. 
Eventually I'd like us to preserve the premultiplication status in the native image, thus we would get rid of m_premultiplyAlpha and especially m_forciblyPremultiplyAlpha

**Review by mattwoodrow — APPROVED** [review]

**Pull request merged**